File: content/chapters/io/lab/support/send-receive/sender.c
// SPDX-License-Identifier: BSD-3-Clause

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "sock-utils.h"

const char *localhost = "127.0.0.1";

int read_input_with_prompt(char *buf)
{
	char *ret;

	printf("Message for receiver (type 'exit' to quit): ");

	memset(buf, 0, BUFSIZE);
	ret = fgets(buf, BUFSIZE, stdin);
	if (!ret)
		return -1;
	buf[strlen(buf)-1] = '\0';

	return 0;
}

int main(void)
{
	char buf[BUFSIZE];
	int ret, sockfd;
	struct sockaddr_in receiver_addr;

	/* Create a socket. */
	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
	DIE(sockfd < 0, "socket");

	/* TODO 1: Add address and port to receiver_addr. */
	/* HINT: use populate_sockaddr(). */

	while (1) {
		ret = read_input_with_prompt(buf);
		DIE(ret < 0, "read_input_with_prompt");

		/* TODO 2: Send message from buf through the socket with sendto(). */
		/* HINT: Use receiver_addr as the fifth arg. You will need to cast it. */
		/* Use DIE to check for errors. */

		if (strncmp(buf, "exit", 4) == 0)
			break;
	}

	/* Close the socket */
	ret = close(sockfd);
	DIE(ret < 0, "close");

	return 0;
}

File: content/io/lab/support/send-receive/sender.py
# SPDX-License-Identifier: BSD-3-Clause

import socket  # Exposes the Berkely Sockets API

IP = "127.0.0.1"  # Loopback IP address (localhost)
PORT = 5000  # Port to send data to (non-privileged ports are > 1023)


def main():
    # Create a socket. Do not mind the arguments for now.
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    while True:
        # Get user input as a string.
        message = input("Message for receiver (type 'exit' to quit): ")

        # Send the message to the server. The message must be encoded to bytes.
        sock.sendto(message.encode(), (IP, PORT))

        # If the message is "exit", break out of the loop.
        if message == "exit":
            break

    sock.close()


if __name__ == "__main__":
    main()
